### PR TITLE
Avoid mmap

### LINKED ISSUE #182: xbps_file_hash_check/xbps_verify_file_signature should work block-wise
These functions work by mmapping the entire file and then calling SHA256_Update once, which is stressful to the memory allocation and fails on 32-bit systems with large packages (e.g. urbanterror-data with 1.8G):

    mmap2(NULL, 1874829312, PROT_READ, MAP_PRIVATE, 5, 0) = -1 ENOMEM (Cannot allocate memory)

Instead, we should read the file in chunks (best using read(2)...) and call SHA256_Update several times.  I guess common code between these functions can be extracted as well.

### PR DESCRIPTION
**Pull request by Gottox**
fixes #182

I left mmap for the *.sig files in which I consider safe.

### CODE REVIEW

File: lib/util_hash.c [hunk 1]
```diff
@@ -108,24 +108,38 @@ xbps_mmap_file(const char *file, void **mmf, size_t *mmflen, size_t *filelen)
...
 }
 
+unsigned char *
+xbps_file_hash_raw(const char *file) {
+	int fd;
+	ssize_t len;
+	unsigned char *res, digest[SHA256_DIGEST_LENGTH], buf[65536];
+	SHA256_CTX sha256;
+
+	if ((fd = open(file, O_RDONLY)) < 0)
```
Comment (dumblob) [inline lib/util_hash.c]: Wouldn't it better to `close()` the `fd` in this function before returning?

Comment (Gottox) [inline lib/util_hash.c]: already fixed :)

File: lib/util_hash.c [hunk 2]
```diff
@@ -108,24 +108,39 @@ xbps_mmap_file(const char *file, void **mmf, size_t *mmflen, size_t *filelen)
...
+		return NULL;
+	SHA256_Init(&sha256);
+	while((len = read(fd, buf, sizeof(buf))) > 0)
+		SHA256_Update(&sha256, buf, len);
+	SHA256_Final(digest, &sha256);
+	close(fd);
+
+	res = malloc(SHA256_DIGEST_LENGTH);
+	memcpy(res, digest, SHA256_DIGEST_LENGTH);
```
Comment (chneukirchen) [inline lib/util_hash.c]: why not store in res directly?

File: lib/verifysig.c [hunk 3]
```diff
@@ -128,16 +122,16 @@ xbps_verify_file_signature(struct xbps_repo *repo, const char *fname)
...
 
 out:
 	if (hexfp)
 		free(hexfp);
 	if (rkeyfile)
 		free(rkeyfile);
-	if (buf)
-		(void)munmap(buf, buflen);
+	if (digest)
+		free(digest);
```
Comment (chneukirchen) [inline lib/verifysig.c]: it's ok to free null pointers btw.

File: lib/util_hash.c [hunk 4]
```diff
@@ -108,24 +108,39 @@ xbps_mmap_file(const char *file, void **mmf, size_t *mmflen, size_t *filelen)
...
+
+	if ((fd = open(file, O_RDONLY)) < 0)
+		return NULL;
+	SHA256_Init(&sha256);
+	while((len = read(fd, buf, sizeof(buf))) > 0)
+		SHA256_Update(&sha256, buf, len);
+	SHA256_Final(digest, &sha256);
+	close(fd);
+
+	res = malloc(SHA256_DIGEST_LENGTH);
```
Comment (chneukirchen) [inline lib/util_hash.c]: unchecked malloc

File: lib/verifysig.c (hunk 3 above)
Comment (Gottox) [inline lib/verifysig.c]: Yea, but these checks are everywhere. I kept it consistent.

**Comment by radare** [comment]
Whats the reasoning behind this change? I guessed i missed some irc

**Comment by Gottox** [comment]
@radare We have packages that fail with oom errors on 32 bit due to the excessive use of mmap.

File: lib/util_hash.c (hunk 2 above)
Comment (Gottox) [inline lib/util_hash.c]: good point, fixing.

**Comment by Gottox** [comment]
Any comments before I merge?

**Comment by radare** [comment]
thanks for the explanation :)

**Pull request merged**